scipy/linalg/flapack_other.pyf.src

```diff
@@ -72,26 +111,138 @@ subroutine <prefix2c>tgexc(wantq,wantz,n,a,lda,b,ldb,q,ldq,z,ldz,ifst,ilst,info)
+    integer intent(hide),depend(b) :: ldb=MAX(1,shape(b,0))
     <ftype2c> intent(in,out,copy),dimension(ldq,n) :: q
-    integer intent(hide),depend(q) :: ldq=shape(q,0)
+    integer intent(hide),depend(q) :: ldq=MAX(1,shape(q,0))
     <ftype2c> intent(in,out,copy),dimension(ldz,n) :: z
-    integer intent(hide),depend(z) :: ldz=shape(z,0)
+    integer intent(hide),depend(z) :: ldz=MAX(1,shape(z,0))
     integer intent(in) :: ifst
     integer intent(in) :: ilst
     integer intent(out) :: info
 
 end subroutine <prefix2c>tgexc
 
+subroutine <prefix2>trsen(job,wantq,select,n,a,lda,q,ldq,wr,wi,m,s,sep,work,lwork,iwork,liwork,info)
+
+    callstatement (*f2py_func)(job,(wantq?"V":"N"),select,&n,a,&lda,q,&ldq,wr,wi,&m,&s,&sep,work,&lwork,iwork,&liwork,&info)
+    callprotoargument char*,char*,F_INT*,F_INT*,<ctype2>*,F_INT*,<ctype2>*,F_INT*,<ctype2>*,<ctype2>*,F_INT*,<ctype2>*,<ctype2>*,<ctype2>*,F_INT*,F_INT*,F_INT*,F_INT*
+
+    character optional,intent(in),check(*job=='N'||*job=='E'||*job=='V'||*job=='B'):: job = 'B'
+    logical optional, intent(in),check(wantq==0||wantq==1) :: wantq=1
+    logical intent(in),dimension(n),depend(n) :: select
+    integer intent(hide),depend(a) :: n=shape(a,0)
+    <ftype2> intent(in,out,copy,out=as),dimension(n,n) :: a
+    integer intent(hide),depend(a) :: lda=MAX(1,shape(a,0))
+    <ftype2> intent(in,out,copy,out=qs),dimension(n,n),depend(n) :: q
+    integer intent(hide),depend(q) :: ldq=MAX(1,shape(q,0))
+    <ftype2> intent(out),dimension(n),depend(n) :: wr
+    <ftype2> intent(out),dimension(n),depend(n) :: wi
+    integer intent(out) :: m
+    <ftype2> intent(out) :: s
+    <ftype2> intent(out) :: sep
+    <ftype2> intent(hide),dimension(MAX(lwork,1)) :: work
+    ! These lwork and liwork values are bare minimum estiamates only for the case job == "N".
+    ! A separate lwork query is a prerequisite due to m dependence.
+    ! Depending on the m value lwork can go up to n**2 / 2 for n = 2m hence it is not
+    ! possible to give a minimal value without a potential excessive memory waste.
+    integer optional,intent(in),depend(n),check(lwork == -1 || lwork >= 1) :: lwork=MAX(1,n)
+    integer intent(hide),dimension(MAX(1,liwork)) :: iwork
+    integer optional,intent(in),check(liwork == -1 || liwork >= 1) :: liwork=1
+    integer intent(out) :: info
+
+end subroutine <prefix2>trsen
+
+subroutine <prefix2>trsen_lwork(job,wantq,select,n,a,lda,q,ldq,wr,wi,m,s,sep,work,lwork,iwork,liwork,info)
+
+    fortranname <prefix2>trsen
+    callstatement (*f2py_func)(job,"N",select,&n,a,&lda,&q,&ldq,&wr,&wi,&m,&s,&sep,&work,&lwork,&iwork,&liwork,&info)
+    callprotoargument char*,char*,F_INT*,F_INT*,<ctype2>*,F_INT*,<ctype2>*,F_INT*,<ctype2>*,<ctype2>*,F_INT*,<ctype2>*,<ctype2>*,<ctype2>*,F_INT*,F_INT*,F_INT*,F_INT*
+
+    character optional,intent(in),check(*job=='N'||*job=='E'||*job=='V'||*job=='B'):: job = 'B'
+    logical intent(in),dimension(n),depend(n) :: select
+    <ftype2> intent(in),dimension(n,n) :: a
```

Comment:
For `lwork` queries, the only things needed are `job`, `compq` and `n`. That avoids having a huge matrix passed to f2py and possibly converted only to be discarded later. This is safe because the unneeded arguments are not referenced and we find those variables by checking the Fortran code directly. 

We do this trick in many places (or at least in the ones we touched in the last years). Here is one example where everything else is dummified to scalars and `work` is shortened https://github.com/scipy/scipy/blob/main/scipy/linalg/flapack_other.pyf.src#L508=

Reply:
I just copy-pasted it from `tgsen_lwork` that you implemented a few years ago: https://github.com/scipy/scipy/commit/b6414d9accda02ab30a3d38b994c5224ede4bd43. I think `trsen` and `tgsen` should use the same calling conventions. Should I also modify `tgsen_lwork`?

Reply:
Yes but in trsen and tgsen the array itself is referenced so we have to pass them. These are not one-size-fits-for-all wrappers unfortunately. LAPACK people do strange things with each routine but probably they have a good reason. The remaining ones shouldn't need those.

Reply:
Sorry, I don't really understand. I only implement an `_lwork` method for `trsen`. `trsen` doesn't seem to reference `a` when `lwork=-1`, but `tgsen` does, because it has to compute `m`. I could make `trsen_lwork` to not require `a`, but then you'd get something like

```
    lwork = _compute_lwork(tgsen_lwork, select, s, t)
```
and
```
    lwork = _compute_lwork(trsen_lwork, select, t.shape[0])
```
instead of
```
    lwork = _compute_lwork(trsen_lwork, select, t)
```
which seems inconsistent.

Reply:
There is no consistency required for these wrappers because this is inconsistent in the LAPACK anyways. If you have a C-contiguous array from numpy and you want to compute the lwork you have to pass the whole array and f2py will happily create another copy of this array then Fortran routine won't even touch it. So we hide as much as we can. No consistency is sought here but mostly convenience and performance. Hence instead of the whole array we use `n` which is in my opinion much simpler.

Reply:
Actually, this won't work. DTRSEN also uses `t`:

https://netlib.org/lapack/explore-html/da/dba/group__double_o_t_h_e_rcomputational_gafe383d8eb4c3815d8946a5e24ba1899a.html

So I think this is fine as is.

Reply:
Trust me it does not 

```fortran

       wantbh = lsame( job, 'B' )
       wants = lsame( job, 'E' ) .OR. wantbh
       wantsp = lsame( job, 'V' ) .OR. wantbh
       wantq = lsame( compq, 'V' )
 *
       info = 0
       lquery = ( lwork.EQ.-1 )
       IF( .NOT.lsame( job, 'N' ) .AND. .NOT.wants .AND. .NOT.wantsp )
      $     THEN
          info = -1
       ELSE IF( .NOT.lsame( compq, 'N' ) .AND. .NOT.wantq ) THEN
          info = -2
       ELSE IF( n.LT.0 ) THEN
          info = -4
       ELSE IF( ldt.LT.max( 1, n ) ) THEN
          info = -6
       ELSE IF( ldq.LT.1 .OR. ( wantq .AND. ldq.LT.n ) ) THEN
          info = -8
       ELSE
 *
 *        Set M to the dimension of the specified invariant subspace,
 *        and test LWORK and LIWORK.
 *
          m = 0
          pair = .false.
          DO 10 k = 1, n
             IF( pair ) THEN
                pair = .false.
             ELSE
                IF( k.LT.n ) THEN
                   IF( t( k+1, k ).EQ.zero ) THEN
                      IF( SELECT( k ) )
      $                  m = m + 1
                   ELSE
                      pair = .true.
                      IF( SELECT( k ) .OR. SELECT( k+1 ) )
      $                  m = m + 2
                   END IF
                ELSE
                   IF( SELECT( n ) )
      $               m = m + 1
                END IF
             END IF
    10    CONTINUE
 *
          n1 = m
          n2 = n - m
          nn = n1*n2
 *
          IF( wantsp ) THEN
             lwmin = max( 1, 2*nn )
             liwmin = max( 1, nn )
          ELSE IF( lsame( job, 'N' ) ) THEN
             lwmin = max( 1, n )
             liwmin = 1
          ELSE IF( lsame job, 'E' ) ) THEN
             lwmin = max( 1, nn )
             liwmin = 1
          END IF
 *
          IF( lwork.LT.lwmin .AND. .NOT.lquery ) THEN
             info = -15
          ELSE IF( liwork.LT.liwmin .AND. .NOT.lquery ) THEN
             info = -17
          END IF
       END IF
 *
       IF( info.EQ.0 ) THEN
          work( 1 ) = lwmin
          iwork( 1 ) = liwmin
       END IF
 *
       IF( info.NE.0 ) THEN
          CALL xerbla( 'DTRSEN', -info )
          RETURN
       ELSE IF( lquery ) THEN
          RETURN
       END IF
```

Reply:
wait why are we changing the variable names ? I was under the impression that a was not referenced but this is a and q now for some reason. Let's not do that please.

Reply:
I don't see your comment here, on github, but did receive it in the mail. `t` is referenced here in the code you copy-pasted:

```
 *
 *        Set M to the dimension of the specified invariant subspace,
 *        and test LWORK and LIWORK.
 *
          m = 0
          pair = .false.
          DO 10 k = 1, n
             IF( pair ) THEN
                pair = .false.
             ELSE
                IF( k.LT.n ) THEN
                   IF( t( k+1, k ).EQ.zero ) THEN
```

and the tests indeed start failing when I don't pass `t`.

Reply:
Yes because I was looking for `a` but somehow it is renamed see my comment above

Reply:
Ah, yes, that's my bad. I did that because of tgsen, but LAPACK names them differently (a and b for tgsen, t for trsen). I'll change it to be t, thanks!